HaxeManual/05-expressions.tex

```diff
@@ -169,11 +169,242 @@ \section{Constants}
+
+\paragraph{Bitwise operators}
+
+\begin{center}
+\begin{tabular}{| l | l | l | l | l |}
+	\hline
+	Operator & Operation & Operand 1 & Operand 2 & Result type \\ \hline
+	\expr{<<} & shift left & \type{Int} & \type{Int} & \type{Int} \\
+	\expr{>>} & shift right & \type{Int} & \type{Int} & \type{Int} \\
+	\expr{>>>} & unsigned shift right & \type{Int} & \type{Int} & \type{Int} \\
+	\expr{\&} & bitwise and & \type{Int} & \type{Int} & \type{Int} \\
+	\expr{|} & bitwise or & \type{Int} & \type{Int} & \type{Int} \\
+	\expr{\^} & bitwise xor & \type{Int} & \type{Int} & \type{Int}
+\end{tabular}
+\end{center}
+
+\paragraph{Logical operators}
+
+\begin{center}
+\begin{tabular}{| l | l | l | l | l |}
+	\hline
+	Operator & Operation & Operand 1 & Operand 2 & Result type \\ \hline
+	\expr{\&\&} & logical and & \type{Bool} & \type{Bool} & \type{Bool} \\
+	\expr{||} & logical or & \type{Bool} & \type{Bool} & \type{Bool}
+\end{tabular}
+\end{center}
+
+\emph{Short-circuiting:}
+
+Haxe guarantees that compound boolean expressions with the same operator are evaluated from left to right and only as far as necessary at run-time. For instance, an expression like \expr{A \&\& B} will evaluate \expr{A} first and evaluate \expr{B} only if the evaluation of \expr{A} yielded \expr{true}. Likewise, the expressions \expr{A || B} will not evaluate \expr{B} if the evaluation of \expr{A} yielded \expr{true}, because the value of \expr{B} is irrelevant in that case. This is important in cases such as this:
+
+\begin{lstlisting}
+if (object != null && object.field == 1) { }
+\end{lstlisting}
+
+Accessing \expr{object.field} if \expr{object} is \expr{null} would lead to a run-time error, but the check for \expr{object != null} guards against it.
+
+\paragraph{Compound assignment operators}
+
+\begin{center}
+\begin{tabular}{| l | l | l | l | l |}
+	\hline
+	Operator & Operation & Operand 1 & Operand 2 & Result type \\ \hline
+	\expr{\%=} & modulo & \type{Float/Int} & \type{Float/Int} & \type{Float/Int} \\
+	\expr{*=} & multiplication & \type{Float/Int} & \type{Float/Int} & \type{Float/Int} \\
+	\expr{/=} & division & \type{Float} & \type{Float/Int} & \type{Float} \\
+	\expr{+=} & addition & \type{Float/Int} & \type{Float/Int} & \type{Float/Int} \\
+	\expr{-=} & subtraction & \type{Float/Int} & \type{Float/Int} & \type{Float/Int} \\
+	\expr{<<=} & shift left & \type{Int} & \type{Int} & \type{Int} \\
+	\expr{>>=} & shift right & \type{Int} & \type{Int} & \type{Int} \\
+	\expr{>>>=} & unsigned shift right & \type{Int} & \type{Int} & \type{Int} \\
+	\expr{\&=} & bitwise and & \type{Int} & \type{Int} & \type{Int} \\
+	\expr{|=} & bitwise or & \type{Int} & \type{Int} & \type{Int} \\
+	\expr{\^=} & bitwise xor & \type{Int} & \type{Int} & \type{Int}
+\end{tabular}
+\end{center}
+
+In all cases, a compound assignment modifies the value of a variable, then returns the new value.
```

Comment:
I know you can interpret fields and arrays as "variable", but I'd still make it clear that we're not strictly talking about `var`s here.

Reply:
Also not sure about the "returns" terminology. I think "evaluates to" is better to avoid potential confusion.

Reply:
Another thing that could be mentioned is that you can't do `Int += Float`, which is different from `Int + Float` where this works just fine.